src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

The implementation of the `WebhookSecretGuard` class looks good in terms of functionality and security. However, there are a few suggestions to improve the code:

1. Consider adding error handling for the cryptographic operations to handle potential exceptions.
2. It's recommended to use a constant-time comparison function for comparing the HMAC digest with the signature to prevent timing attacks.

```diff
+ // Add error handling for cryptographic operations
+ try {
+     const hmac = crypto.createHmac('sha256', secret);
+     const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+ } catch (error) {
+     throw new UnauthorizedException('Error in cryptographic operation');
+ }
+
+ // Use constant-time comparison function
+ if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+     throw new UnauthorizedException('Invalid signature');
+ }
```

This will enhance the robustness and security of the guard implementation.



<!-- This is an auto-generated comment by PullProofAI -->